misk/src/main/kotlin/misk/flags/DynamicFlagStore.kt

```diff
@@ -0,0 +1,22 @@
+package misk.flags
+
+import java.util.concurrent.TimeUnit
+
+/**
+ * Applications register their interest in monitoring flags through one of the registerXXXFlag
+ * calls. If registration and watching requires contacting a remote service (e.g. a flag service,
+ * Zookeeper, etc), this should be done asynchronously so that multiple flags can be registered
+ * in parallel, with the XXXFlag.get() calls blocking until the registration for that flag is
+ * complete. An application can also call [awaitRegistrationsComplete] to block until all
+ * pending registrations have completed, allowing e.g. the application to avoid processing
+ * incoming requests until all of its flags are available and watched
+ */
+interface DynamicFlagStore {
```

Comment:
Are all flags strings behind-the-scenes? Or does registration bind the name to its type?

It’s weird that for booleans the store manages how those are mapped to persistence
But for JSONs that mapping happens in the caller

I think it might be simpler for this to be strings only?

Reply:
I think it's reasonable to allow the underlying backend to support optimized storage for common types (strings, numbers, booleans), and layer support for derived types (JSON, etc) on top of those base types.